test/contrib/beam_dataflow_test.py

```diff
@@ -0,0 +1,261 @@
+            '--runner=DirectRunner',
+            '--input=some-input-dir/part-*',
+            '--output=some-output.txt'
+        ]
+
+        self.assertEqual(cmd_line_args, expected)
+
+    def test_dataflow_full_cmd_line_args(self):
+        task = FullTestTask()
+        cmd_line_args = task._mk_cmd_line()
+
+        expected = [
+            'java',
+            'com.spotify.luigi.SomeJobClass',
+            '--runner=DirectRunner',
+            '--project=some-project',
+            '--zone=europe-west1-c',
+            '--region=europe-west1',
+            '--stagingLocation=some-staging',
+            '--tempLocation=some-temp',
+            '--gcpTempLocation=some-gcp-temp',
+            '--numWorkers=1',
+            '--autoscalingAlgorithm=THROUGHPUT_BASED',
+            '--maxNumWorkers=2',
+            '--diskSizeGb=5',
+            '--workerMachineType=n1-standard-4',
+            '--workerDiskType=compute.googleapis.com/projects//zones//diskTypes/pd-ssd',
+            '--network=some-network',
+            '--subnetwork=some-subnetwork',
+            '--jobName=SomeJobName',
+            '--serviceAccount=some-service-account@google.com',
+            '--labels={"k1": "v1"}',
+            '--extraArg=present',
+            '--input=some-input-dir/part-*',
+            '--output=some-output.txt'
+        ]
+
+        self.assertEqual(json.loads(cmd_line_args[19][9:]), {'k1': 'v1'})
+        self.assertEqual(cmd_line_args, expected)
+        self.assertEqual(task.output_uris, {"output": "some-output.txt"})
+
+    def test_dataflow_with_file_patterns(self):
+        cmd_line_args = FilePatternsTestTask()._mk_cmd_line()
+
+        self.assertIn('--input1=some-input-dir/part-*', cmd_line_args)
+        self.assertIn('--input2=some-input-dir/*.some-ext', cmd_line_args)
+
+    def test_dataflow_with_invalid_file_patterns(self):
+        task = FilePatternsTestTask()
+        task.file_pattern = MagicMock(return_value='notadict')
+        with self.assertRaises(ValueError):
+            task._mk_cmd_line()
+
+    def test_dataflow_runner_resolution(self):
+        task = SimpleTestTask()
+        # Test that supported runners are passed through
+        for runner in ["DirectRunner", "BlockingDataflowPipelineRunner",
+                       "InProcessPipelineRunner", "DataflowRunner"]:
+            task.runner = runner
+            self.assertEqual(task._get_runner(), runner)
+
+        # Test that unsupported runners default to DirectRunner
+        task.runner = "UnsupportedRunner"
+        self.assertEqual(task._get_runner(), "DirectRunner")
+
+    def test_dataflow_successful_run_callbacks(self):
+        task = DummyCmdLineTestTask()
+
+        task.before_run = MagicMock()
+        task.validate_output = MagicMock()
+        task.on_successful_run = MagicMock()
+        task.on_output_validation = MagicMock()
+        task.cleanup_on_error = MagicMock()
+
+        task.run()
+
+        task.before_run.assert_called_once_with()
```

Comment:
Just `assert_called_once()` should be enough https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once

Reply:
Hm, I used `assert_called_once_with` because it's also a method that's also available in the Python 2 unittest library (`assert_called_once` isn't). It seems like other unit test in the luigi project are doing the same thing

Reply:
Thanks. Good to know.